### PR TITLE
Compare annotations before discarding CiliumNode updates.

### PR DESCRIPTION
**Pull request by LynneD**
Currently we discard CiliumNode updates based on DeepEqual and labels. However DeepEqual is set to ignore Annotations, and the wg-pub-key annotation is used to exchange rotated Wireguard keys.

### CODE REVIEW

**Comment by LynneD** [comment]
Hi @aojea, this is the fix for Wiregurad tunnel drop issue. 

The wireguard tunnel is broken when any node restart happens. After restart, the restarted node's public key got refreshed but not propagated to other nodes.

The issue is caused by cilium dropping CiliumNode update events when the spec, status and labels between the old and new nodes are the same. Wireguard public key updates are transmitted through annotations. 

We want to confirm if this is a known issue since it can be reproduced by restarting any node easily.

Thanks!

File: pkg/k8s/watchers/cilium_node.go
Comment (aojea) [inline pkg/k8s/watchers/cilium_node.go:74]: This causes issue in case the wireguard key changes but the rest of the node object not
https://github.com/cilium/cilium/blob/e52fe1d59d1c6f51a76993641ea9412e0f44f749/pkg/nodediscovery/nodediscovery.go#L460-L465

since, we don't compare Annotations the event is dropped and not processed

Comment (tommyp1ckles) [inline pkg/k8s/watchers/cilium_node.go:74]: It might be unclear for someone reading this why we need to compare annotations - could you add a comment explaining the context?

**Comment by tommyp1ckles** [comment]
Thanks for the PR 😄 

I wonder if instead of checking all annotation changes, it might make more sense to target the specific wireguard key annotation and just see if that's changed.

It seems like the intention is to avoid unnecessary node updates.

**Comment by aojea** [comment]
> I wonder if instead of checking all annotation changes, it might make more sense to target the specific wireguard key annotation and just see if that's changed.

in the long term my experience is that new annotations got added but you never remember to update this comparison, it is also coherent with existing comparison with labels

File: pkg/k8s/watchers/cilium_node.go
Comment (LynneD) [reply pkg/k8s/watchers/cilium_node.go:74]: Updated the commit with comment.

**Comment by aojea** [comment]
The failed jobs in "cilium conformance e2e" show 

> ssh: connect to host localhost port 2222: Connection refused

**Review by gandro — APPROVED** [review]
Looks good to me. I also prefer checking all annotations as the current PR does, since I do believe that we will hit similar bugs in the future otherwise.

**Comment by danehans** [comment]
I think I see a similar issue with [pod annotations](https://github.com/cilium/cilium/blob/52942f6682510eefaca34f3c1b1b5db0c005712f/pkg/k8s/watchers/pod.go#L299-L305). The pod controller is comparing the [EgressBandwidth annotation](https://pkg.go.dev/github.com/cilium/cilium/pkg/bandwidth) but not the [IngressBandwidth](https://pkg.go.dev/github.com/cilium/cilium/pkg/bandwidth) annotation.

**Comment by LynneD** [comment]
Travis CI - Pull Request build error shows:

`apt-get install failed
The command "sudo -E apt-get -yq --no-install-suggests --no-install-recommends $(travis_apt_get_options) install kernel-package gnupg libncurses5" failed and exited with 100 during .`

Is there a way to retrigger the build?

**Comment by danehans** [comment]
@LynneD and others, I'm seeing the [same](https://github.com/cilium/cilium/pull/25465#issuecomment-1550026025) build failure in my PRs too ([xref](https://app.travis-ci.com/github/cilium/cilium/builds/263087167)).

**Comment by christarazi** [comment]
Thanks for the fix. Could you format your commit msg in the way described by the contribution [docs](https://docs.cilium.io/en/v1.13/contributing/development/contributing_guide/#submit-pr)? See step 5.

**Comment by LynneD** [comment]
> Thanks for the fix. Could you format your commit msg in the way described by the contribution [docs](https://docs.cilium.io/en/v1.13/contributing/development/contributing_guide/#submit-pr)? See step 5.

Thanks @christarazi! Do I need to create a issue for this commit? I don't have a issue number currently.

**Comment by christarazi** [comment]
No need for an issue at this time, thanks!

**Comment by aojea** [comment]
/test

failed on "wait for images to be available"

**Comment by christarazi** [comment]
Reopened PR to retrigger GH actions.

**Comment by christarazi** [comment]
@LynneD The checkpatch job is failing because it is complaining about the sign-off in the commit. I believe it wants the sign-off to contain: 

```
Lin Dong <lindongchn@gmail.com>
```

Btw, this can be done automatically by running `git commit --sign-off`.

**Comment by LynneD** [comment]
> @LynneD The checkpatch job is failing because it is complaining about the sign-off in the commit. I believe it wants the sign-off to contain:
> 
> ```
> Lin Dong <lindongchn@gmail.com>
> ```
> 
> Btw, this can be done automatically by running `git commit --sign-off`.

Updated.

**Comment by danehans** [comment]
@christarazi thanks for the review. I squashed the commits, PTAL when you have a moment.

**Review by brb — CHANGES_REQUESTED** [review]
Good finds! Could you add https://github.com/cilium/cilium/pull/25465#issuecomment-1548605569 to the commit message?

**Comment by LynneD** [comment]
> Good finds! Could you add [#25465 (comment)](https://github.com/cilium/cilium/pull/25465#issuecomment-1548605569) to the commit message?

Added the comment.

**Comment by christarazi** [comment]
> > Good finds! Could you add [#25465 (comment)](https://github.com/cilium/cilium/pull/25465#issuecomment-1548605569) to the commit message?
> 
> Added the comment. Is the the PR ready to merge? We are waiting for the CL to be merged to cherry-pick it to our internal branch.

A PR is ready to merge when all codeowners have approved the PR and the CI is all green. Right now, Martynas is required to approve (I've re-requested him) and I will trigger the CI.

**Comment by LynneD** [comment]
> > > Good finds! Could you add [#25465 (comment)](https://github.com/cilium/cilium/pull/25465#issuecomment-1548605569) to the commit message?
> > 
> > 
> > Added the comment. Is the the PR ready to merge? We are waiting for the CL to be merged to cherry-pick it to our internal branch.
> 
> A PR is ready to merge when all codeowners have approved the PR and the CI is all green. Right now, Martynas is required to approve (I've re-requested him) and I will trigger the CI.

Thanks @christarazi

**Review by tommyp1ckles — APPROVED** [review]
> I agree with @aojea's point here. Is there a specific concern about checking all annotation changes?

Nothing specific - mostly just wondering if it would lead to unnecessary updates. This seems reasonable to me.

**Comment by LynneD** [comment]
> Good finds! Could you add [#25465 (comment)](https://github.com/cilium/cilium/pull/25465#issuecomment-1548605569) to the commit message?

Hi @brb, I've updated the commit message, can you please take another look?

**Review by brb — APPROVED** [review]
Thanks!

**Comment by LynneD** [comment]
Thanks @christarazi @brb @tommyp1ckles @aojea, I've got enough approvals, and the "k8s-1.26-kernel-net-next" test seems not related to this PR. Can you please take another look and see how I can unblock the merging?

**Comment by christarazi** [comment]
I've rebased to re-run the tests. Please don't push unless the CI points out a failure and you need to push to address it.

**Comment by LynneD** [comment]
> I've rebased to re-run the tests. Please don't push unless the CI points out a failure and you need to push to address it.

Got it. Thanks @christarazi!

**Comment by christarazi** [comment]
Thanks for the PR and your patience!

**Comment by LynneD** [comment]
Thanks so much @christarazi!

**Pull request merged**
